### PR TITLE
Documentation build cleanup

### PR DESCRIPTION
**Pull request by hodgestar**
Our documentation currently isn't building on readthedocs and it's producing quite a few warnings. This PR intends to clean all that up.

Issues addressed so far:

- [x] Add double underscores to anonymous external references to prevent RST warnings.
- [x] Reference the non-existent debug mode documentation consistently throughout the docs.
- [x] Give runtime and helper API documentation different section names to prevent a Sphinx warning.
- [x] Update removal of HPyAPI_HELPER define to help spinx-c-autodoc / clang properly find these functions.
- [x] Remove attempt to call cindex.Index.create that segfaults in some cases.

Still to do:

- [x] Debug the failing [readthedocs builds](https://readthedocs.org/projects/hpy/builds/) (hopefully easier now that less is going wrong)
- [x] Discuss the complete removal of the call to `cindex.Index.create`
- [x] Discuss making the build fail in the are warnings produced while running the documentation build (this would have picked up some of the issues that resulted in our build missing documentation or failing)

### CODE REVIEW

File: docs/conf.py
Comment (hodgestar) [inline docs/conf.py:87]: @antocuni Are you okay with me just deleting this call `cindex.Index.create()` now? It segfaults consistently on my machine. Not sure if your setup still needs it or not though.

**Review by fangerer — APPROVED** [review]
thanks !

File: docs/conf.py
Comment (antocuni) [reply docs/conf.py:87]: it segfaults also for me, I'm fine with deleting.
But maybe keep a comment saying that on ubuntu you need `apt install libclang-10-dev`?

**Comment by hodgestar** [comment]
It looks like sorting out the other issues (probably the segfault on Ubuntu) fixed the readthedocs builds. Woot.

**Comment by hodgestar** [comment]
@fangerer @antocuni (or anyone else): Ready for a proper review.

I added some very basic documentation of the debug mode to make the documentation build pass without warnings.

File: azure-pipelines.yml
Comment (antocuni) [inline azure-pipelines.yml:195]: if I understand it correctly, the point of this job is just to check that the docs build, and then we discard the result? If so, maybe add a comment to explain that.
Also, note that as soon as we merge #230 we probably we need re-add this job to the CI

File: docs/debug-mode.rst
Comment (antocuni) [inline docs/debug-mode.rst:5]:
```suggestion
HPy includes a debug mode which includes a lot of useful run-time checks to
ensure that C extensions use the API correctly. The major points of the debug mode are:

    1. no special compilation flags are required: it is enough to compile the extension 
       with the Universal ABI.
    
    2. The debug mode can be activated at *import time*, and it can be activated
       per-extension.
    
    3. You pay the overhead of the debug mode only if you use it. Extensions loaded 
       without the debug mode run at full speed.

This is possible because the whole of the HPy API is provided
```

File: docs/misc/str-builder-api.rst
Comment (antocuni) [inline docs/misc/str-builder-api.rst:230]: just for curiosity, what is the difference between a single and a double `__`?

**Review by antocuni — APPROVED** [review]
LGTM, thanks!

**Comment by antocuni** [comment]
> See, for example, https://docs.hpyproject.org/en/feature-doc-build-cleanup/api-reference/index.html

it's a 404. Maybe we need to enable some readthedocs setting to build the docs also for PRs?

File: docs/misc/str-builder-api.rst
Comment (hodgestar) [reply docs/misc/str-builder-api.rst:230]: A single `_` automatically creates a reference with the name of the link, which is a problem if multiple links have the same text. The double `__` prevents the reference being created.

File: azure-pipelines.yml
Comment (hodgestar) [reply azure-pipelines.yml:195]: Correct. I will add a comment. When we add it back to GitHub actions it might be nice so that one can download the docs from the build for debugging purposes.

I already noted the existence of this PR in #230.

**Comment by hodgestar** [comment]
> it's a 404. Maybe we need to enable some readthedocs setting to build the docs also for PRs?

Meh. I deactivated the automated readthedocs builds for this branch once I knew they were working. I didn't realize that would also make the URL not work.

File: azure-pipelines.yml
Comment (hodgestar) [reply azure-pipelines.yml:199]: Comment added to explain things.

**Comment by hodgestar** [comment]
Just waiting for the CI pipelines to finish and then I'll merge in the morning. Merging should also update the readthedocs documentation (which is a bit behind because the builds were failing there for awhile).

**Pull request merged**
